python/grass/temporal/temporal_granularity.py

```diff
@@ -344,164 +423,83 @@ def compute_absolute_time_granularity(maps):
-                dlist.append(d["minute"])
-            elif "hour" in d and d["hour"] > 0:
-                dlist.append(d["hour"] * 60)
-            elif "day" in d:
-                dlist.append(d["day"] * 24 * 60)
-            else:
-                dlist.append(d["max_days"] * 24 * 60)
-        use_minutes = True
-    elif has_hours:
-        for d in datetime_delta:
-            if "hour" in d and d["hour"] > 0:
-                dlist.append(d["hour"])
-            elif "day" in d and d["day"] > 0:
-                dlist.append(d["day"] * 24)
-            else:
-                dlist.append(d["max_days"] * 24)
-        use_hours = True
-    elif has_days:
-        for d in datetime_delta:
-            if "day" in d and d["day"] > 0:
-                dlist.append(d["day"])
-            else:
-                dlist.append(d["max_days"])
-        use_days = True
-    elif has_months:
-        for d in datetime_delta:
-            if "month" in d and d["month"] > 0:
-                dlist.append(d["month"])
-            elif "year" in d and d["year"] > 0:
-                dlist.append(d["year"] * 12)
-        use_months = True
-    elif has_years:
-        for d in datetime_delta:
-            if "year" in d:
-                dlist.append(d["year"])
-        use_years = True
-
-    dlist.sort()
-    ulist = list(set(dlist))
-
-    if len(ulist) == 0:
+    dlist = set()
+    for time_unit, granularity_set in granularity_units.items():
+        # The smallest granularity unit is used so as soon as a non-zero
+        # value / set is encountered, the loop can be aborted
+        if granularity_set:
+            dlist.update(granularity_set)
+            break
+
+    if not dlist:
         return None
 
-    if len(ulist) > 1:
+    if len(dlist) > 1:
         # Find greatest common divisor
-        granularity = gcd_list(ulist)
+        granularity = gcd_list(dlist)
     else:
-        granularity = ulist[0]
+        granularity = dlist.pop()
 
-    if use_seconds:
-        if granularity == 1:
-            return "%i second" % granularity
-        else:
-            return "%i seconds" % granularity
-    elif use_minutes:
-        if granularity == 1:
-            return "%i minute" % granularity
-        else:
-            return "%i minutes" % granularity
-    elif use_hours:
-        if granularity == 1:
-            return "%i hour" % granularity
-        else:
-            return "%i hours" % granularity
-    elif use_days:
-        if granularity == 1:
-            return "%i day" % granularity
-        else:
-            return "%i days" % granularity
-    elif use_months:
-        if granularity == 1:
-            return "%i month" % granularity
-        else:
-            return "%i months" % granularity
-    elif use_years:
-        if granularity == 1:
-            return "%i year" % granularity
-        else:
-            return "%i years" % granularity
+    if granularity is None or time_unit is None:
```

Comment:
In an IDE with pyright, it says time_unit is possibly unbound. And effectively it is always used as a loop variable, never really defined.

Reply:
Right, changed in in [04d4778](https://github.com/OSGeo/grass/pull/3350/commits/04d4778d32e34a4c3842187fd22bb7fbc4a8d34e)